flux-jobs: color completed, failed, cancelled jobs

chu11 opened the pull request:
Built upon PR #2910 and utilizing the "result" attribute,  colorize "completed" and "failed" jobs green/red accordingly.  It's a super simple patch (just the last one in this commit stream).

![Screenshot from 2020-05-13 16-23-00](https://user-images.githubusercontent.com/274859/81876220-fe926100-9536-11ea-9eba-3a6bab2cece8.png)

- Added a `--color` option which takes `never` or `auto` as args, modeled it after the `ls --color` option.

- I decided not to color cancelled jobs, maybe I should?  I figure cancelled jobs, the user cancelled for a reason and knows.  They could be red?  Or another color?

- It'd be easy to support a configuration (environment variable or option) where we have rules to say "completed=34;failed=36" or whatever, where users can select color via the ansi code.  Easy to parse and easy to check inputs and then use in escape codes.  Feels like too much for now, given I'm only colorizing two outputs, but perhaps later.

- Contemplated only coloring the jobs of the user running the `flux-jobs` command, but not sure if that's a good idea given we will be supporting job submissions by guest users.  Perhaps of whatever user is passed in via `--user`?  But that might be confusing to an average user, why some failed jobs aren't being colored??  So I decided not to do it.

grondo commented:
> I decided not to color cancelled jobs, maybe I should? I figure cancelled jobs, the user cancelled for a reason and knows. They could be red? Or another color?

Perhaps cancelled jobs could be greyed out, since presumably they didn't run.

> Contemplated only coloring the jobs of the user running the flux-jobs command, but not sure if that's a good idea given we will be supporting job submissions by guest users.

I agree it would be confusing if displayed jobs aren't all treated the same. If anything, you could have an option to make jobs of other users shaded a bit darker. This might not make a good default for a sysadmin view (presumably they'd have no jobs running), but perhaps we'd have a different tool for that.

>  Perhaps of whatever user is passed in via --user? But that might be confusing to an average user, why some failed jobs aren't being colored?? So I decided not to do it.

If `--user` is used, won't that user's jobs be the only one's displayed anyway?

chu11 commented:
> Perhaps cancelled jobs could be greyed out, since presumably they didn't run.

Oh that's a pretty good idea.  Lets try that.

![Screenshot from 2020-05-14 11-06-52](https://user-images.githubusercontent.com/274859/81969597-11f30a00-95d3-11ea-80c0-f43b25fd02e5.png)

Looks pretty good!  Some tables listed that grey-ish color as "white", ugh ... didn't think of it as an option.

> If --user is used, won't that user's jobs be the only one's displayed anyway?

Oh duh! :-)

chu11 commented:
rebased just so there aren't conflicts

chu11 commented:
rebased now that #2910 is in and removed WIP.

grondo commented:
Maybe just my terminal, but how does it look for you with the colors changed to bold instead of normal? e.g.
```diff
diff --git a/src/cmd/flux-jobs.py b/src/cmd/flux-jobs.py
index 5dee43944..51e0e3a22 100755
--- a/src/cmd/flux-jobs.py
+++ b/src/cmd/flux-jobs.py
@@ -593,9 +593,9 @@ def color_setup(args, job):
     if args.color == "auto" and sys.stdout.isatty():
         if job.result:
             if job.result == "COMPLETED":
-                sys.stdout.write("\033[32m")
+                sys.stdout.write("\033[01;32m")
             elif job.result == "FAILED":
-                sys.stdout.write("\033[31m")
+                sys.stdout.write("\033[01;31m")
             elif job.result == "CANCELLED":
                 sys.stdout.write("\033[37m")
```

This matches the colors used by `ls`:
```console
 grondo@asp:~/git/flux-core.git$ ls --color=always /bin/ls | cat -v
^[[0m^[[01;32m/bin/ls^[[0m
```

The regular blue and red are difficult to read on my terminal.

chu11 commented:
oh i should mention, i didn't add any tests ... i can't seem to find that correct magic to get this done.  `script` can capture color output, but haven't figured out the right magic to try and "grep" from it.  I think I can grep the trailing `0m`, but that's all i figured out so far.  Dunno if anyone has any tricks they know of.

garlick commented:
1990's me is slightly horrified by hardwiring VT-100 escape sequences here, but I regularly suppress that guy so no worries.

Is there some python module that would at least provide the mapping of named colors to codes (and maybe do nothing if not on the right sort of terminal?)

chu11 commented:
> Maybe just my terminal, but how does it look for you with the colors changed to bold instead of normal? 

Looks pretty good to me.

![Screenshot from 2020-05-19 13-36-33](https://user-images.githubusercontent.com/274859/82375740-dd69be80-99d5-11ea-94bf-7c2bf316bb58.png)

the latter is the original, bolded on top.

The bolded green is my default for executables in `ls`, so I think I prefer a different green.  But I will submit to the popular opinion :-)   Or perhaps should just the red be bolded?

grondo commented:
> Dunno if anyone has any tricks they know of.

Do you need to ensure `flux-jobs` runs under a tty? If so, use the `t/scripts/runpty.py` command will do this for you, and log output in a couple formats (raw, asciicast, or eventlog, though not sure if that is helpful)

The log is encoded in utf-8, so this seems to work:

```console
 $ t/scripts/runpty.py -f asciicast flux jobs -n 2577450139648 | grep '\u001b\[37m'
```

grondo commented:
> The bolded green is my default for executables in ls, so I think I prefer a different green.

Perhaps `ls` chose bolded green because the normal green is not easily readable on some terminals? (I may be missing why `flux-jobs` output colors need to differ from `ls` though..)

chu11 commented:
> Is there some python module that would at least provide the mapping of named colors to codes (and maybe do nothing if not on the right sort of terminal?)

there are some modules, but none standard (to my knowledge).  The 'colored' and 'colorama' modules seem popular, and apparently there's a python36 `colorama` RPM for redhat.  So maybe that one is worth checking out.  Let me look into it.

chu11 commented:
> Perhaps ls chose bolded green because the normal green is not easily readable on some terminals? (I may be missing why flux-jobs output colors need to differ from ls though..)

I think it's just a personal preference. @garlick or @SteVwonder, bold or not bold?

garlick commented:
Probably bold is more to my liking, but I don't have a strong opinion.

Regarding colorama etc, I wouldn't want to hold this up when we already have it working.  Maybe something to ponder for a future PR?

grondo commented:
> Regarding colorama etc, I wouldn't want to hold this up when we already have it working. Maybe something to ponder for a future PR?

Not sure we really want to have yet another `pip` requirement for something simple like this. The standard `curses` module for python has some color information, but I'm not sure if you need to `initscr()` before using it (which we definitely *don't* want to do).

Eventually we might require some other curses package like `urwid` or `blessings` and then we can switch to using its color settings since we'll have a new, more significant dependency.

grondo commented:
> oh i should mention, i didn't add any tests .

One simple test that we should have is to ensure that no color output is generated when stdout is not a tty:

```
flux jobs -a > output &&
test_must_fail grep '^[' output
```

Note to type literal escape character type Ctrl-V Esc

chu11 commented:
re-pushed

- completed & failed jobs are now bolded green & red (cancelled grey color still the same)
- the `--color` option now takes `auto` (default), `never`, and `always` just like `ls`.  The `always` was mostly for testing purposes.
- added a manpage entry for `--color`
- added bunch-o-tests

note the diff may look weird b/c of the ESC character in the grep.

chu11 commented:
Hmmm so this test is often failing ... first time it like most of the builders, then just one of them (after my fixups).  Can't imagine this is racy.  Trying some debug commits.

```
not ok 45 - flux-jobs --color=auto works (pty)
FAIL: t2800-jobs-cmd.t 45 - flux-jobs --color=auto works (pty)
#	
#	        $runpty flux jobs --suppress-header --color=auto --filter=pending | tail -n +2 > colorP_auto.out &&
#	        check_no_color colorP_auto.out &&
#	        $runpty flux jobs --suppress-header --color=auto --filter=running | tail -n +2 > colorR_auto.out &&
#	        check_no_color colorR_auto.out &&
#	        $runpty flux jobs --suppress-header --color=auto --filter=completed | tail -n +2 > colorCD_auto.out &&
#	        count=`grep "\\u001b\[01\;32m" colorCD_auto.out | wc -l` &&
#	        test $count -eq 4 &&
#	        $runpty flux jobs --suppress-header --color=auto --filter=failed | tail -n +2 > colorF_auto.out &&
#	        count=`grep "\\u001b\[01\;31m" colorF_auto.out | wc -l` &&
#	        test $count -eq 1 &&
#	        $runpty flux jobs --suppress-header --color=auto --filter=cancelled | tail -n +2 > colorCA_auto.out &&
#	        count=`grep "\\u001b\[37m" colorCA_auto.out | wc -l` &&
#	        test $count -eq 1
#	
```

grondo commented:
> Can't imagine this is racy. Trying some debug commits.

One thing to note about `runpty` if using either `eventlog` or `asciicast` output formats -- the script is reading unbuffered output, so the number of lines/json objects in the output is not predictable from run to run (necessarily).

(Not sure if that is the problem you are seeing, but i wanted to point it out.)

chu11 commented:
> One thing to note about runpty if using either eventlog or asciicast output formats -- the script is reading unbuffered output, so the number of lines/json objects in the output is not predictable from run to run (necessarily).

Your guess was correct.  The number of lines was short (two lines merged into one).  So I guess I gotta count occurrences instead of just grep.

```
[0.12735438346862793, "o", "\u001b[01;32m       29074915328 travis   hostname       CD      1      1   0.080s 0\r\n"]
[0.12748265266418457, "o", "\u001b[0;0m\u001b[01;32m       25031606272 travis   hostname       CD      1      1   0.101s 0"]
[0.1275196075439453, "o", "\r\n"]
[0.1279582977294922, "o", "\u001b[0;0m\u001b[01;32m       20669530112 travis   hostname       CD      1      1   0.103s 0\r\n\u001b[0;0m\u001b[01;32m       15988686848 travis   hostname       CD      1      1   0.115s 0\r\n\u001b[0;0m"]
```

chu11 commented:
woo hoo, travis passed using `grep -o` to fix the counting problem.  I re-pushed removing those debug commits, but left the fixups in there.

grondo commented:
Looks like you just have to remove the fixup commit. Then I say set MWP!

chu11 commented:
cool, squashed and re-pushed, with a minor edit to one commit message.
